phpseclib/Crypt/RSA.php

```diff
@@ -2072,55 +2055,133 @@ function _rsassa_pkcs1_v1_5_verify($m, $s)
+        }
+
+        // RSA verification
+
+        $s = $this->_os2ip($s);
+        $m2 = $this->_rsavp1($s);
+        if ($m2 === false) {
+            return false;
+        }
+        $em = $this->_i2osp($m2, $this->k);
+        if ($em === false) {
+            return false;
+        }
+
+        if ($this->_string_shift($em, 2) != "\0\1") {
+            return false;
+        }
+
+        $em = ltrim($em, "\xFF");
+        if ($this->_string_shift($em) != "\0") {
+            return false;
+        }
+
+        $asn1 = new ASN1();
+        $decoded = $asn1->decodeBER($em);
+        if (!is_array($decoded) || empty($decoded[0]) || strlen($em) > $decoded[0]['length']) {
+            return false;
+        }
+
+        $AlgorithmIdentifier = array(
+            'type'     => ASN1::TYPE_SEQUENCE,
+            'children' => array(
+                'algorithm'  => array('type' => ASN1::TYPE_OBJECT_IDENTIFIER),
+                'parameters' => array(
+                                    'type'     => ASN1::TYPE_ANY,
+                                    'optional' => true
+                                )
+            )
+        );
+
+        $DigestInfo = array(
+            'type'     => ASN1::TYPE_SEQUENCE,
+            'children' => array(
+                'digestAlgorithm' => $AlgorithmIdentifier,
+                'digest' => array('type' => ASN1::TYPE_OCTET_STRING)
+            )
+        );
+
+        $oids = array(
+            '1.2.840.113549.2.2' => 'md2',
+            '1.2.840.113549.2.4' => 'md4', // from PKCS1 v1.5
+            '1.2.840.113549.2.5' => 'md5',
+            '1.3.14.3.2.26' => 'sha1',
+            '2.16.840.1.101.3.4.2.1' => 'sha256',
+            '2.16.840.1.101.3.4.2.2' => 'sha384',
+            '2.16.840.1.101.3.4.2.3' => 'sha512',
+            // from PKCS1 v2.2
+            //'2.16.840.1.101.3.4.2.5' => 'sha512/224',
+            //'2.16.840.1.101.3.4.2.6' => 'sha512/256',
+        );
+
+        $asn1->loadOIDs($oids);
+
+        $decoded = $asn1->asn1map($decoded[0], $DigestInfo);
+        if (!isset($decoded) || $decoded === false) {
+            return false;
+        }
+
+        if (!in_array($decoded['digestAlgorithm']['algorithm'], $oids)) {
+            return false;
+        }
+
+        $hash = new Hash($decoded['digestAlgorithm']['algorithm']);
+        $em = $hash->hash($m);
+        $em2 = base64_decode($decoded['digest']);
+
+        return $this->_equals($em, $em2);
     }
 
     /**
      * Encryption
      *
-     * Both self::ENCRYPTION_OAEP and self::ENCRYPTION_PKCS1 both place limits on how long $plaintext can be.
+     * Both self::PADDING_OAEP and self::PADDING_PKCS1 both place limits on how long $plaintext can be.
      * If $plaintext exceeds those limits it will be broken up so that it does and the resultant ciphertext's will
      * be concatenated together.
      *
      * @see self::decrypt()
      * @access public
      * @param string $plaintext
+     * @param int $padding
```

Comment:
Should this be `@param int $padding optional`, as seems to be the convention for optional parameters?